### PR TITLE
added ability to set a hostgroup by matching the hostname with regex

### PR DESCRIPTION
**Pull request by lotherk**
This adds the functionallity to add hosts to hostgroups matching their hostname on a regex pattern.

There is currently only a simple yaml file to create the matching map, this should be moved in the hostgroup section or something.

I think the best place would be if we would add some new fields into the hostgroup UI like in the puppetclasses where you can add match values.

### CODE REVIEW

File: lib/default_hostgroup_managed_host_patch.rb [hunk 1]
```diff
@@ -4,12 +4,48 @@ def self.included(base)
...
 
   module ClassMethods
+
+    def importHostAndFacts_with_match_hostgroup hostname, facts, certname = nil, proxy_id = nil
+      host, result = importHostAndFacts_without_match_hostgroup(hostname, facts, certname, proxy_id)
+      Rails.logger.debug "DefaultHostgroupMatch: performing Hostgroup match"
+      return host, result if host.hostgroup
+
+      begin
+        _map = YAML.load_file("hostgroupmatch.yml")
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Any particular reason why you're starting all your variables with an underscore?

Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: I think I'd like to see this move into a SETTINGS file (in ~foreman/config/settings.plugins.d) - see https://github.com/GregSutcliffe/foreman_column_view/blob/master/app/helpers/foreman_column_view/hosts_helper_extension.rb#L13 for an example of how to do this (or ping me on IRC)

Comment (lotherk) [inline lib/default_hostgroup_managed_host_patch.rb]: Not really, it's more or less a (bad?) habit from another language. I use method params without _ and variables inside methods with _. You're free to change this and I'm trying to change my habit. ;-)

Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: I dislike changing other peoples commits on merge, I might accidentally break something. Please remove them, just from a style viewpoint :)

File: lib/default_hostgroup_managed_host_patch.rb [hunk 2]
```diff
@@ -4,12 +4,48 @@ def self.included(base)
...
+        _hostgroup = nil
+
+        _map.each do |regex, hostgroup|
+
+          unless valid_hostgroup?(hostgroup)
+            Rails.logger.error "DefaultHostgroupMatch: #{hostgroup} (#{regex}) is no valid hostgroup, skipping."
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Use Foreman::Exception here. Also there's a typo, it whould be "... is not a valid hostgroup... "

File: lib/default_hostgroup_managed_host_patch.rb [hunk 3]
```diff
@@ -29,8 +65,8 @@ def importHostAndFacts_with_apply_hostgroup hostname, facts, certname = nil, pro
       return host, result
     end
 
-    def valid_hostgroup?
-      Hostgroup.find_by_label(Setting[:default_hostgroup]) ? true : false
+    def valid_hostgroup?(hostgroup = Setting[:default_hostgroup])
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Does it make sense to even have Setting[:default_hostgroup] now? Users can just add a final "/.*/ => Foo" line in their map for a global catchall, right?

Comment (lotherk) [inline lib/default_hostgroup_managed_host_patch.rb]: Yes, that's right.

Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Cool, any chance you can drop the Setting model in your patch then? :)

Comment (lotherk) [inline lib/default_hostgroup_managed_host_patch.rb]: Of course. :-)

**Comment by GregSutcliffe** [comment]
@lotherk this is good stuff, just a few small tweaks and we can get it merged. Moving to including the map in SETTINGS is probably the messiest part, ping me if you need a hand.

**Comment by GregSutcliffe** [comment]
@lotherk oh, we'll need tests too - again, I can help if you need a hand.

File: lib/default_hostgroup_managed_host_patch.rb [hunk 4]
```diff
@@ -4,12 +4,48 @@ def self.included(base)
     base.extend ClassMethods
     base.class_eval do
       class << self
+        alias_method_chain :importHostAndFacts, :match_hostgroup  # before :apply_hostgroup
         alias_method_chain :importHostAndFacts, :apply_hostgroup
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: I think we can drop the old method entirely - if people want a global default, they can add it to the map.

**Comment by lotherk** [comment]
@GregSutcliffe I've added tests and added force_hostgroup_match to the settings page to force matching on hosts that already have a hostgroup. default is false. I've also added tests. Would be good if you'd take a look on them. :-)

File: app/models/setting/default_hostgroup.rb [hunk 5]
```diff
@@ -7,7 +7,8 @@ def self.load_defaults
 
       Setting.transaction do
         [
-          self.set('default_hostgroup', 'The default Hostgroup to place new Hosts in', ''),
+          self.set('force_hostgroup_match', 'Apply hostgroup matching even if a host already has one.', false),
+          self.set('force_hostgroup_match_only_new', 'Apply hostgroup matching only on new hosts', true),
```
Comment (GregSutcliffe) [inline app/models/setting/default_hostgroup.rb:11]: Isn't this the same thing? Why do we need two settings to decide how to handle a host with/without a hostgroup?

File: lib/default_hostgroup_managed_host_patch.rb [hunk 6]
```diff
@@ -4,33 +4,53 @@ def self.included(base)
...
-      if host.present? && !host.new_record? && host.hostgroup.nil? && host.reports.empty?
-        host.hostgroup = Hostgroup.find_by_label(Setting[:default_hostgroup])
-        host.save(:validate => false)
-        Rails.logger.debug "DefaultHostgroup: added #{host.name} to #{host.hostgroup.label}"
+      map = SETTINGS[:default_hostgroup][:map]
+      new_hostgroup = nil
+
+      map.each do |hostgroup, regex|
+        unless valid_hostgroup?(hostgroup)
+          Rails.logger.error "DefaultHostgroupMatch: #{hostgroup} (#{regex}) is not a valid hostgroup, skipping."
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: nitpick: Why mention the regex in the exception? you're only testing the hostgroup here

File: lib/default_hostgroup_managed_host_patch.rb [hunk 7]
```diff
@@ -4,33 +4,53 @@ def self.included(base)
...
+          next
+        end
+        regex.gsub!(/(^\/|\/$)/, '')
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb]: Use \A and \z here instead of ^ and $

File: app/models/setting/default_hostgroup.rb (hunk 5 above)
Comment (GregSutcliffe) [inline app/models/setting/default_hostgroup.rb:11]: Scratch that, read the README :)

**Comment by GregSutcliffe** [comment]
@lotherk Looking good. There's a lot of duplicated code in the tests that we can probably extract to a helper, but otherwise this is shaping up nicely.

I'm off to FOSDEM in a couple of hours, so I'll pick this up with you next week somethime :)

File: lib/default_hostgroup_managed_host_patch.rb (hunk 6 above)
Comment (lotherk) [inline lib/default_hostgroup_managed_host_patch.rb]: yes, indeed. :)

File: lib/default_hostgroup_managed_host_patch.rb (hunk 7 above)
Comment (lotherk) [inline lib/default_hostgroup_managed_host_patch.rb]: Thanks for that hint!

**Comment by GregSutcliffe** [comment]
@lotherk ok, conference season is over, let's pick this up. I'll try to take a look at the current state in the next day or two :)

**Comment by lotherk** [comment]
NOW the commits are squashed. ;-)

**Comment by GregSutcliffe** [comment]
@lotherk I totally forgot about this, sorry! I'll try to find some time to retest it this week.

**Comment by lotherk** [comment]
haha, no problem. ;-)

**Comment by GregSutcliffe** [comment]
I've just created a develop branch to match my other plugins - any chance you could rebase against that? The only change is we have to use find_by_title for Hostgroups now.

I'm testing the actual patch now, comments to follow :)

File: lib/default_hostgroup_managed_host_patch.rb [hunk 8]
```diff
@@ -1,36 +1,55 @@
...
-      unless valid_hostgroup?
-        Rails.logger.debug "DefaultHostgroup: Could not find Hostgroup '#{Setting[:default_hostgroup]}'"
-        return host, result
+      Rails.logger.debug "DefaultHostgroupMatch: performing Hostgroup match"
+      host, result = importHostAndFacts_without_match_hostgroup(hostname, facts, certname, proxy_id)
+
+      if Setting[:force_hostgroup_match_only_new]
+        return host, result unless host.present? && !host.new_record? && host.hostgroup.nil? && host.reports.empty?
+      end
+
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb:22]: You can write this as

``` ruby
return host, result unless !Setting[:force_hostgroup_match_only_new] && host.present? && !host.new_record? && host.hostgroup.nil? && host.reports.empty?
```

File: lib/default_hostgroup_managed_host_patch.rb [hunk 9]
```diff
@@ -1,36 +1,55 @@
...
+      unless Setting[:force_hostgroup_match]
+        return host, result if host.hostgroup
       end
```
Comment (GregSutcliffe) [inline lib/default_hostgroup_managed_host_patch.rb:25]: Likewise here, I think

``` ruby
return host, result if host.hostgroup && !Setting[:force_hostgroup_match]
```

is more readable.

**Comment by GregSutcliffe** [comment]
Works great! Tests pass too :)

Maybe I missed it, but I don't see a way to say <clear hostgroup> for a given regex - i.e unset the hostgroup. Maybe we could use a special string? Say

```
:map
  "-nil-" => /^.*\.blankhosts\.com$/
```

What do you think?

**Comment by GregSutcliffe** [comment]
Merged as 1418c02c7494e0c64bbfac4e8236c9cdcecb4a3e - thanks for the hard work!

**Pull request closed without merging**